nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -2007,6 +2057,265 @@ impl DataStore {
+                        old_volume
+                    } else {
+                        // Existing volume was deleted, so return an error. We
+                        // can't perform the snapshot replacement now!
+                        return Err(err.bail(
+                            VolumeReplaceSnapshotError::TargetVolumeDeleted
+                        ));
+                    };
+
+                    let old_vcr: VolumeConstructionRequest =
+                        match serde_json::from_str(&old_volume.data()) {
+                            Ok(vcr) => vcr,
+                            Err(e) => {
+                                return Err(err.bail(
+                                    VolumeReplaceSnapshotError::SerdeError(e)
+                                ));
+                            },
+                        };
+
+                    // Does it look like this replacement already happened?
+                    let old_target_in_vcr = match target_in_vcr(&old_vcr, &existing) {
+                        Ok(v) => v,
+                        Err(e) => {
+                            return Err(err.bail(
+                                VolumeReplaceSnapshotError::SnapshotReplacementError(e)
+                            ));
+                        },
+                    };
+
+                    let new_target_in_vcr = match target_in_vcr(&old_vcr, &replacement) {
+                        Ok(v) => v,
+                        Err(e) => {
+                            return Err(err.bail(
+                                VolumeReplaceSnapshotError::SnapshotReplacementError(e)
+                            ));
+                        },
+                    };
+
+                    if !old_target_in_vcr && new_target_in_vcr {
+                        // It does seem like the replacement happened
+                        return Ok(());
+                    }
+
+                    // Update the existing volume's construction request to
+                    // replace the existing target's SocketAddrV6 with the
+                    // replacement target's
+
+                    // Copy the old volume's VCR, changing out the old target
+                    // for the new.
+                    let (new_vcr, replacements) = match replace_read_only_target_in_vcr(
+                        &old_vcr,
+                        existing,
+                        replacement,
+                    ) {
+                        Ok(new_vcr) => new_vcr,
+                        Err(e) => {
+                            return Err(err.bail(
+                                VolumeReplaceSnapshotError::SnapshotReplacementError(e)
+                            ));
+                        }
+                    };
+
+                    // Expect that this only happened once. If it happened
+                    // multiple times, question everything: how would a snapshot
+                    // be used twice?!
```

Comment:
Yeah, this would be a mess.  I almost want to panic here, as what the hell
is going on in our volume?  Should we even allow this to happen?

What if the `replace_read_only_target_in_vcr()` just stopped when it
got to the first match?  Would that be better, or worse?

Reply:
I don't really understand what is going on here, but I wouldn't panic because it would cause this nexus and maybe others to all be shut down. This type of thing should go into our pile of alerts that is being designed now.

Reply:
If this happens it means we have a VCR that has the same target (the I.P. address of one of the downstairs) in
more than one location.  It means we have done something terribly wrong somewhere and
this VCR is invalid.

There is no good that can come from this.. but.. to walk back my desire to panic.
If we did arise at this situation, I think crucible may catch errors and whatever instance was
trying to use the volume will get stuck for most of the ways it could manifest.

If we have two targets in the same region set that was writeable, then crucible it will never activate 
because a downstairs will only ever allow one RW activation at a time.

If we have two targets the same between a RW region set and a RO region set (the read_only_parent),
then crucible will refuse to attach to one of them as crucible checks to see if a downstairs is RW/RO
and won't connect if that downstairs is different than what it expects.

If we have two targets the same in the read_only_parent, I'm not sure what happens here.
Worst case would be that it is allowed to move forward and one downstairs answers IO requests twice
from the same upstairs, but the upstairs thinks it's talking to two different downstairs.

Best case would be the downstairs refuses to connect, but I need to go read the code.

At the end of the day though, we have somehow constructed an invalid VCR and we need to 
know how that happened.  If we just log a message and move on, I fear that message would never be
seen.

Reply:
Okay, I dug into the downstairs code, and I think we are "safe" if we send crucible a
bad VCR (one with the same downstairs target repeated).  The RO code has this check:
```
           // Multiple active read-only sessions are allowed, but multiple     
            // sessions for the same Upstairs UUID are not. Kick out a          
            // previously active session for this UUID if one exists.
```
So, this protects us from harm.  While a duplicated targets is a sign of a serious problem, it
should not result in corruption and the VCR will not "work", so whatever damage is done
will be limited.

Reply:
In terms of constructing invalid VCRs that would trip this, I think this would only happen if we could "concatenate" the higher level disk objects somehow. Then one could construct two disks from the same source (a snapshot or image), then "join" them together to make one larger disk. That larger disk's volume would have two sub volumes, both of which are volumes that have a read-only parent that contains the target we're trying to replace. It's arguable that the _correct_ thing to do in this case is replace both targets, but this will currently be rejected by the following logic in `compare_vcr_for_replacement`:

```rust
                // Presently, we only support one sub_volume for replacement.
                // If support for multiple sub_volumes is added, then this
                // following section will need to be updated to loop over the
                // sub_volume Vec and find the specific one with a difference,
                // while verifying that all other sub_volumes are no different.
                if n_sub_volumes.len() != 1 {
                    crucible_bail!(
                        ReplaceRequestInvalid,
                        "Only a single sub_volume is supported"
                    )
                }
```

More broadly, this sort of concatenation operation doesn't make sense and I don't see why we would build such a thing. Some sort of "grow" operation that would add a sub volume of blank blocks makes sense to do, but not a concatenation.